src/realm/sync/instruction_applier.cpp

```diff
@@ -848,6 +848,78 @@ void InstructionApplier::resolve_list(const Instruction::PathInstruction& instr,
     resolve_path(instr, instr_name, std::move(callback));
 }
 
+bool InstructionApplier::allows_null_links(const Instruction::PathInstruction& instr, const char* instr_name)
+{
+    bool allows_nulls = false;
+    auto callback = util::overload{[&](LstBase&, size_t) {},
+                                   [&](LstBase&) {},
+                                   [&](SetBase&) {},
+                                   [&](Dictionary&) {
+                                       allows_nulls = true;
+                                   },
+                                   [&](Dictionary&, Mixed) {
+                                       allows_nulls = true;
+                                   },
+                                   [&](Obj&, ColKey) {
+                                       allows_nulls = true;
+                                   }};
+    resolve_path(instr, instr_name, std::move(callback));
+    return allows_nulls;
+}
+
+std::string InstructionApplier::print(const Instruction::PathInstruction& instr) const
+{
+    REALM_ASSERT(m_log);
+    std::stringstream ss;
+    m_log->print_path(ss, instr.table, instr.object, instr.field, &instr.path);
+    return ss.str();
+}
+
+bool InstructionApplier::check_links_exist(const Instruction::Payload& payload)
+{
+    bool valid_payload = true;
+    using Type = Instruction::Payload::Type;
+    if (payload.type == Type::Link) {
+        StringData class_name = get_string(payload.data.link.target_table);
+        Group::TableNameBuffer buffer;
+        StringData target_table_name = Group::class_name_to_table_name(class_name, buffer);
+        TableRef target_table = m_transaction.get_table(target_table_name);
+        if (!target_table) {
+            bad_transaction_log("Link with invalid target table '%1'", target_table_name);
+        }
+        if (target_table->is_embedded()) {
+            bad_transaction_log("Link to embedded table '%1'", target_table_name);
+        }
+        Mixed linked_pk = mpark::visit(
+            util::overload{
```

Comment:
I agree that your suggestion is cleaner code. I think of this kind of like a switch statement with a default case. My preference is to be explicit about all types so that if we add another primary key type we will get a compiler error rather than some hard to find bug due to implicit casting or something. However I'm not stuck on that, so if you think the smaller code is  better let me know and I'll change it.